### DISCUSSION TITLE
Example showing interface with Julia

### DISCUSSION THREAD

**Discussion by ben18785** [pints-team/pints#1647] [Ideas/Science]
Julia has really fast ODE solvers. I'd like to have an example notebook where we show how Julia solvers can be used from within Python, so that they can be used with PINTS. Perhaps even also show how PINTS can be called from Julia. It seems like there are some fairly mature packages for this interfacing [here](https://github.com/JuliaPy/pyjulia).

**Comment by MichaelClerx** [comment]
Solving ODEs is outside the scope of PINTS. So if anyone wants to code up those examples that's fine and can go in separate repos (similar to https://github.com/pints-team/myokit-pints-examples ).

**Comment by ben18785** [comment]
Ok but can we point to an example of this within PINTS? I agree solving ODEs is outside of scope but given that increasing numbers of people solve ODEs within Julia, I think it’d be good to show them that PINTS can be used in this way.

On that note, it’d be good to perhaps have a repo for miscellaneous stuff related to fitting ODE/PDE/DAEs with PINTS. In these, we could show how to, fit example, calculate sensitivities using Sundials/CVODE and use these in PINTS. In this, we could show how to interface Julia/other software with PINTS too.

**Comment by MichaelClerx** [comment]
Just do a repo for each! So feel free to set up a pints/team-julia-ode or something :-)

**Comment by Lethay** [comment]
Hey, did you get any further with this? I tried using [DiffEqPy](https://github.com/SciML/diffeqpy) to solve a system with Julia's ODE solver, because the ODE solvers do indeed perform much better than SciPy's for the problem I'm currently solving. However, the problem I ran into is that Python and Julia don't play nice in a multiprocessing context and it runs excruciatingly slowly. I suspect that Julia is just-in-time compiling every iteration, not even merely once per core. In my example, what would take up to 20 seconds in pure python instead took up to 10 minutes with the Julia integration, sadly.

This problem is a [known issue](https://pyjulia.readthedocs.io/en/latest/limitations.html) but I was interested to see if you'd made any progress regardless.

It was an unfortunate first foray into Julia, to find that it'd be a choice between good data fitting in python or good ODE solving in Julia!
